{"context": " More than 1 year has passed since last update.import Network\n   (withSocketsDo, connectTo, PortID(PortNumber))\n\nimport System.IO\n   (hGetContents)\n\nimport System.Environment\n   (getArgs)\n\nimport Control.Exception\n   (catch, SomeException)\n\n-- Prelude \u306e catch \u304c Control.Exception \u306e catch \u3068\n-- \u304b\u3076\u3063\u3066\u3066\u90aa\u9b54\u306a\u306e\u3067\u96a0\u3059\u3002\nimport Prelude hiding\n   (catch)\n\nmain :: IO ()\nmain = withSocketsDo main'\n\nmain' :: IO ()\nmain' = do\n hostname <- getHostname\n dayTime  <- getDayTime hostname `catch` errMessage\n putStrLn dayTime\n\ngetHostname :: IO String\ngetHostname = do\n args <- getArgs\n return $ args ?| \"time.nist.gov\"\n\n-- safe head\ninfixr 2 ?|\n(?|) :: [a] -> a -> a\n(?|) (x:_) _ = x\n(?|) _     y = y\n\ndayTimePort :: PortID\ndayTimePort = PortNumber 13\n\ngetDayTime :: String -> IO String\ngetDayTime hostname = do\n handle <- connectTo hostname dayTimePort\n hGetContents handle\n\nerrMessage :: SomeException -> IO String\nerrMessage e = return \"Network error!\"\n\n```hs\nimport Network\n   (withSocketsDo, connectTo, PortID(PortNumber))\n\nimport System.IO\n   (hGetContents)\n\nimport System.Environment\n   (getArgs)\n\nimport Control.Exception\n   (catch, SomeException)\n\n-- Prelude \u306e catch \u304c Control.Exception \u306e catch \u3068\n-- \u304b\u3076\u3063\u3066\u3066\u90aa\u9b54\u306a\u306e\u3067\u96a0\u3059\u3002\nimport Prelude hiding\n   (catch)\n\nmain :: IO ()\nmain = withSocketsDo main'\n\nmain' :: IO ()\nmain' = do\n hostname <- getHostname\n dayTime  <- getDayTime hostname `catch` errMessage\n putStrLn dayTime\n\ngetHostname :: IO String\ngetHostname = do\n args <- getArgs\n return $ args ?| \"time.nist.gov\"\n\n-- safe head\ninfixr 2 ?|\n(?|) :: [a] -> a -> a\n(?|) (x:_) _ = x\n(?|) _     y = y\n\ndayTimePort :: PortID\ndayTimePort = PortNumber 13\n\ngetDayTime :: String -> IO String\ngetDayTime hostname = do\n handle <- connectTo hostname dayTimePort\n hGetContents handle\n\nerrMessage :: SomeException -> IO String\nerrMessage e = return \"Network error!\"\n```", "tags": ["Haskell", "socket"]}